File: Skydown App/Services/Music/ArtistPagesStore.swift
import Foundation
import Combine
import FirebaseFirestore

@MainActor
protocol ArtistPagesServicing {
    func observePages(_ onChange: @escaping @MainActor (Result<[ArtistPage], Error>) -> Void) -> () -> Void
    func save(page: ArtistPage) async throws
}

final class FirebaseArtistPagesService: ArtistPagesServicing {
    private let firestore: Firestore
    private let collectionName = "artistPages"

    init(firestore: Firestore = Firestore.firestore()) {
        self.firestore = firestore
    }

    func observePages(_ onChange: @escaping @MainActor (Result<[ArtistPage], Error>) -> Void) -> () -> Void {
        let listener = firestore.collection(collectionName).addSnapshotListener { snapshot, error in
            Task { @MainActor in
                if let error {
                    onChange(.failure(error))
                    return
                }

                let pages = (snapshot?.documents ?? [])
                    .compactMap(Self.mapPage(document:))
                onChange(.success(pages))
            }
        }

        return {
            listener.remove()
        }
    }

    func save(page: ArtistPage) async throws {
        let payload: [String: Any] = [
            "slug": artistPageSlug(from: page.artistName),
            "brand": page.brand.rawValue,
            "artistName": page.artistName.trimmingCharacters(in: .whitespacesAndNewlines),
            "tagline": page.tagline?.trimmingCharacters(in: .whitespacesAndNewlines) ?? NSNull(),
            "bio": page.bio?.trimmingCharacters(in: .whitespacesAndNewlines) ?? NSNull(),
            "profileImageURL": page.profileImageURL?.trimmingCharacters(in: .whitespacesAndNewlines) ?? NSNull(),
            "heroImageURL": page.heroImageURL?.trimmingCharacters(in: .whitespacesAndNewlines) ?? NSNull(),
            "heroVideoURL": page.heroVideoURL?.trimmingCharacters(in: .whitespacesAndNewlines) ?? NSNull(),
            "instagramURL": page.instagramURL?.trimmingCharacters(in: .whitespacesAndNewlines) ?? NSNull(),
            "spotifyURL": page.spotifyURL?.trimmingCharacters(in: .whitespacesAndNewlines) ?? NSNull(),
            "youtubeURL": page.youtubeURL?.trimmingCharacters(in: .whitespacesAndNewlines) ?? NSNull(),
            "studioPriceList": page.studioPriceList
                .map { item in
                    [
                        "title": item.title.trimmingCharacters(in: .whitespacesAndNewlines),
                        "detail": item.detail.trimmingCharacters(in: .whitespacesAndNewlines),
                        "price": item.price.trimmingCharacters(in: .whitespacesAndNewlines)
                    ]
                }
                .filter { entry in
                    (entry["title"]?.isEmpty == false)
                    && (entry["detail"]?.isEmpty == false)
                    && (entry["price"]?.isEmpty == false)
                },
            "editorUids": Array(Set(page.editorUids.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }.filter { !$0.isEmpty })).sorted(),
            "createdAt": Timestamp(date: page.createdAt),
            "updatedAt": FieldValue.serverTimestamp()
        ]

        let documentID = artistPageDocumentID(brand: page.brand, artistName: page.artistName)
        try await firestore.collection(collectionName).document(documentID).setData(payload, merge: true)
    }

    private static func firstURLString(_ data: [String: Any], _ keys: String...) -> String? {
        for key in keys {
            if let s = (data[key] as? String)?.trimmedNilIfEmpty {
                return s
            }
        }
        return nil
    }

    private static func firstURLStringInNestedMaps(
        _ data: [String: Any],
        mapNames: [String],
        keys: [String]
    ) -> String? {
        for mapName in mapNames {
            guard let m = data[mapName] as? [String: Any] else { continue }
            for k in keys {
                if let s = (m[k] as? String)?.trimmedNilIfEmpty {
                    return s
                }
            }
        }
        return nil
    }

    private static func mapPage(document: QueryDocumentSnapshot) -> ArtistPage? {
        let data = document.data()
        let brandKey = (data["brand"] as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased() ?? ""
        guard
            !brandKey.isEmpty,
            let brand = ArtistPageBrand(rawValue: brandKey),
            let fieldArtistName = data["artistName"] as? String
        else {
            return nil
        }

        let trimmed = fieldArtistName.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        // Firestore `artistName` can be copy-paste wrong; document ID (same as on save) is source of truth.
        let artistName: String
        if brand == .nicma {
            switch document.documentID {
            case "nicma-nicma-music": artistName = "NICMA MUSIC"
            case "nicma-nicma-studio": artistName = "NICMA STUDIO"
            default: artistName = trimmed
            }
        } else {
            artistName = trimmed
        }

        let createdAt = (data["createdAt"] as? Timestamp)?.dateValue() ?? .now
        let updatedAt = (data["updatedAt"] as? Timestamp)?.dateValue() ?? createdAt
        let editorUids = (data["editorUids"] as? [String] ?? [])
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        let studioPriceList = (data["studioPriceList"] as? [[String: Any]] ?? [])
            .compactMap { entry -> StudioPriceItem? in
                guard
                    let title = (entry["title"] as? String)?.trimmedNilIfEmpty,
                    let detail = (entry["detail"] as? String)?.trimmedNilIfEmpty,
                    let price = (entry["price"] as? String)?.trimmedNilIfEmpty
                else {
                    return nil
                }
                return StudioPriceItem(title: title, detail: detail, price: price)
            }

        return ArtistPage(
            id: artistPageDocumentID(brand: brand, artistName: artistName),
            brand: brand,
            artistName: artistName,
            tagline: (data["tagline"] as? String)?.trimmedNilIfEmpty,
            bio: (data["bio"] as? String)?.trimmedNilIfEmpty,
            profileImageURL: (data["profileImageURL"] as? String)?.trimmedNilIfEmpty,
            heroImageURL: (data["heroImageURL"] as? String)?.trimmedNilIfEmpty,
            heroVideoURL: (data["heroVideoURL"] as? String)?.trimmedNilIfEmpty,
            instagramURL: firstURLString(
                data,
                "instagramURL", "instagramUrl", "instagram", "instagramURLString"
            ) ?? firstURLStringInNestedMaps(
                data,
                mapNames: ["social", "links", "link", "instagram"],
                keys: ["instagram", "instagramURL", "instagramUrl", "url", "link", "href", "u"]
            ),
            spotifyURL: firstURLString(data, "spotifyURL", "spotifyUrl", "spotify", "spotifyUrlString")
                ?? firstURLStringInNestedMaps(
                    data,
                    mapNames: ["social", "links", "link", "spotify"],
                    keys: ["spotify", "spotifyURL", "spotifyUrl", "url", "link", "href", "u"]
                ),
            youtubeURL: firstURLString(
                data,
                "youtubeURL", "youtubeUrl", "youtube", "youtubeURLString", "youTubeURL", "youTube"
            ) ?? firstURLStringInNestedMaps(
                data,
                mapNames: ["social", "links", "link", "youtube"],
                keys: ["youtube", "youtubeURL", "youtubeUrl", "yt", "youTube", "url", "link", "href", "u"]
            ),
            studioPriceList: studioPriceList,
            editorUids: editorUids,
            createdAt: createdAt,
            updatedAt: updatedAt,
            isPlaceholder: false
        )
    }
}

final class UITestArtistPagesService: ArtistPagesServicing {
    private var pages: [ArtistPage] = []
    private var observers: [UUID: @MainActor (Result<[ArtistPage], Error>) -> Void] = [:]

    func observePages(_ onChange: @escaping @MainActor (Result<[ArtistPage], Error>) -> Void) -> () -> Void {
        let id = UUID()
        observers[id] = onChange
        onChange(.success(pages))
        return { [weak self] in
            Task { @MainActor in
                self?.observers.removeValue(forKey: id)
            }
        }
    }

    func save(page: ArtistPage) async throws {
        if let index = pages.firstIndex(where: { $0.slug == page.slug && $0.brand == page.brand }) {
            pages[index] = page
        } else {
            pages.append(page)
        }
        notifyObservers()
    }

    private func notifyObservers() {
        for observer in observers.values {
            observer(.success(pages))
        }
    }
}

@MainActor
final class ArtistPagesStore: ObservableObject {
    static let shared = ArtistPagesStore()

    @Published private(set) var pages: [ArtistPage] = []
    @Published private(set) var lastErrorMessage: String?

    private let service: ArtistPagesServicing
    private var stopObserving: (() -> Void)?

    init(service: ArtistPagesServicing? = nil) {
        self.service = service ?? (UITestRuntime.usesIsolatedAuthService ? UITestArtistPagesService() : FirebaseArtistPagesService())
        startObserving()
    }

    func pages(for brand: ArtistPageBrand) -> [ArtistPage] {
        pages.filter { $0.brand == brand }
            .sorted { $0.artistName.localizedCaseInsensitiveCompare($1.artistName) == .orderedAscending }
    }

    func page(for brand: ArtistPageBrand, artistName: String) -> ArtistPage {
        let slug = artistPageDocumentID(brand: brand, artistName: artistName)
        if let page = pages.first(where: { $0.brand == brand && $0.slug == slug }) {
            return page
        }
        return ArtistPage.draft(brand: brand, artistName: artistName)
    }

    func canEdit(_ page: ArtistPage, user: User?) -> Bool {
        guard let user else { return false }
        if user.isPlatformOwner {
            return true
        }

        guard let userId = user.id, !userId.isEmpty else { return false }
        return page.editorUids.contains(userId)
    }

    func save(_ page: ArtistPage) async throws {
        try await service.save(page: page)
    }

    private func startObserving() {
        stopObserving?()
        stopObserving = service.observePages { [weak self] result in
            guard let store = self else { return }

            switch result {
            case .success(let remotePages):
                let merged = store.merge(remotePages: remotePages)
                store.pages = merged
                store.lastErrorMessage = nil
            case .failure(let error):
                store.lastErrorMessage = error.localizedDescription
            }
        }
    }

    private func merge(remotePages: [ArtistPage]) -> [ArtistPage] {
        var pagesByID: [String: ArtistPage] = [:]

        for page in remotePages {
            if let existing = pagesByID[page.slug] {
                pagesByID[page.slug] = resolvedPreferredPage(between: existing, and: page)
            } else {
                pagesByID[page.slug] = page
            }
        }

        return pagesByID.values.sorted { lhs, rhs in
            if lhs.brand != rhs.brand {
                return lhs.brand.rawValue < rhs.brand.rawValue
            }
            return lhs.artistName.localizedCaseInsensitiveCompare(rhs.artistName) == .orderedAscending
        }
    }

    private func resolvedPreferredPage(between lhs: ArtistPage, and rhs: ArtistPage) -> ArtistPage {
        if lhs.updatedAt != rhs.updatedAt {
            return lhs.updatedAt > rhs.updatedAt ? lhs : rhs
        }
        if lhs.isPlaceholder != rhs.isPlaceholder {
            return lhs.isPlaceholder ? rhs : lhs
        }
        return lhs
    }
}

private extension String {
    var trimmedNilIfEmpty: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}

File: androidApp/src/main/java/com/nash/skyos/data/ArtistPagesStore.kt
package com.nash.skyos.data

import com.google.firebase.Timestamp
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.SetOptions
import com.skydown.shared.model.User
import com.skydown.shared.model.isPlatformOwner
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.tasks.await

enum class ArtistPageBrand(val rawValue: String) {
    Zweizwei("zweizwei"),
    Skydown("skydown"),
    Nicma("nicma"),
    ;

    val displayTitle: String
        get() = when (this) {
            Zweizwei -> "22"
            Skydown -> "Skydown"
            Nicma -> "Nicma"
        }
}

data class ArtistPageUi(
    val slug: String,
    val brand: ArtistPageBrand,
    val artistName: String,
    val tagline: String? = null,
    val bio: String? = null,
    val profileImageURL: String? = null,
    val heroImageURL: String? = null,
    val heroVideoURL: String? = null,
    val instagramURL: String? = null,
    val spotifyURL: String? = null,
    val youtubeURL: String? = null,
    val studioPriceList: List<StudioPriceItemUi> = emptyList(),
    val editorUids: List<String> = emptyList(),
    val createdAtEpochMillis: Long = System.currentTimeMillis(),
    val updatedAtEpochMillis: Long = System.currentTimeMillis(),
    val isPlaceholder: Boolean = false,
) {
    val hasCustomPresentation: Boolean
        get() = listOf(
            tagline,
            bio,
            profileImageURL,
            heroImageURL,
            heroVideoURL,
            instagramURL,
            spotifyURL,
            youtubeURL,
            if (studioPriceList.isEmpty()) null else "studio-prices",
        ).any { !it.isNullOrBlank() }
}

data class StudioPriceItemUi(
    val title: String,
    val detail: String,
    val price: String,
)

private fun artistPageSlug(artistName: String): String {
    val normalized = artistName
        .trim()
        .lowercase()
        .replace(Regex("[^a-z0-9]+"), "-")
        .trim('-')

    return normalized.ifBlank { "artist" }
}

private fun artistPageDocumentId(brand: ArtistPageBrand, artistName: String): String {
    return "${brand.rawValue}-${artistPageSlug(artistName)}"
}

/**
 * Firestore-Feld `artistName` ist manchmal falsch kopiert (z. B. beide "NICMA MUSIC");
 * dann würde daraus ein falscher [ArtistPageUi.slug] gebaut und `pageFor("NICMA STUDIO")` findet das Doc nicht.
 * Die **Document-ID** (wie beim [save]) ist die verlässliche Quelle.
 */
private fun canonicalArtistNameForDocument(
    brand: ArtistPageBrand,
    documentId: String,
    storedArtistName: String,
): String {
    if (brand != ArtistPageBrand.Nicma) {
        return storedArtistName
    }
    return when (documentId) {
        "nicma-nicma-music" -> "NICMA MUSIC"
        "nicma-nicma-studio" -> "NICMA STUDIO"
        else -> storedArtistName
    }
}

private fun draftArtistPage(
    brand: ArtistPageBrand,
    artistName: String,
    now: Long = System.currentTimeMillis(),
): ArtistPageUi {
    return ArtistPageUi(
        slug = artistPageDocumentId(brand, artistName),
        brand = brand,
        artistName = artistName,
        createdAtEpochMillis = now,
        updatedAtEpochMillis = now,
        isPlaceholder = true,
    )
}

object ArtistPagesStore {
    private val firestore: FirebaseFirestore
        get() = FirebaseFirestore.getInstance()
    private val collection
        get() = firestore.collection("artistPages")

    private val _pages = MutableStateFlow<List<ArtistPageUi>>(emptyList())
    val pages: StateFlow<List<ArtistPageUi>> = _pages.asStateFlow()

    private val _lastErrorMessage = MutableStateFlow<String?>(null)
    val lastErrorMessage: StateFlow<String?> = _lastErrorMessage.asStateFlow()

    private var listener: ListenerRegistration? = null

    init {
        startObservation()
    }

    fun pagesForBrand(brand: ArtistPageBrand): List<ArtistPageUi> {
        return pages.value
            .filter { it.brand == brand }
            .sortedBy { it.artistName.lowercase() }
    }

    fun pageFor(brand: ArtistPageBrand, artistName: String): ArtistPageUi {
        val slug = artistPageDocumentId(brand, artistName)
        return pages.value.firstOrNull { it.brand == brand && it.slug == slug }
            ?: draftArtistPage(brand = brand, artistName = artistName)
    }

    fun documentIdFor(brand: ArtistPageBrand, artistName: String): String =
        artistPageDocumentId(brand, artistName)

    fun canEdit(page: ArtistPageUi, user: User?): Boolean {
        if (user?.isPlatformOwner == true) {
            return true
        }

        val userId = user?.id?.takeIf { it.isNotBlank() } ?: return false
        return page.editorUids.contains(userId)
    }

    suspend fun save(page: ArtistPageUi): Result<Unit> {
        return runCatching {
            val documentId = artistPageDocumentId(page.brand, page.artistName)
            collection.document(documentId).set(
                mapOf(
                    "slug" to artistPageSlug(page.artistName),
                    "brand" to page.brand.rawValue,
                    "artistName" to page.artistName.trim(),
                    "tagline" to page.tagline.trimmedOrNull(),
                    "bio" to page.bio.trimmedOrNull(),
                    "profileImageURL" to page.profileImageURL.trimmedOrNull(),
                    "heroImageURL" to page.heroImageURL.trimmedOrNull(),
                    "heroVideoURL" to page.heroVideoURL.trimmedOrNull(),
                    "instagramURL" to page.instagramURL.trimmedOrNull(),
                    "spotifyURL" to page.spotifyURL.trimmedOrNull(),
                    "youtubeURL" to page.youtubeURL.trimmedOrNull(),
                    "studioPriceList" to page.studioPriceList
                        .mapNotNull { item ->
                            val title = item.title.trimmedOrNull() ?: return@mapNotNull null
                            val detail = item.detail.trimmedOrNull() ?: return@mapNotNull null
                            val price = item.price.trimmedOrNull() ?: return@mapNotNull null
                            mapOf(
                                "title" to title,
                                "detail" to detail,
                                "price" to price,
                            )
                        },
                    "editorUids" to page.editorUids
                        .mapNotNull { it.trimmedOrNull() }
                        .distinct()
                        .sorted(),
                    "createdAt" to Timestamp(page.createdAtEpochMillis / 1000, 0),
                    "updatedAt" to Timestamp.now(),
                ),
                SetOptions.merge(),
            ).await()
            syncMirroredNicmaSocialLinksIfNeeded(page)
        }
    }

    private suspend fun syncMirroredNicmaSocialLinksIfNeeded(page: ArtistPageUi) {
        if (page.brand != ArtistPageBrand.Nicma) return
        val normalizedArtistName = page.artistName.trim().lowercase()
        val targetArtistName = when (normalizedArtistName) {
            "nicma music" -> "NICMA STUDIO"
            "nicma studio" -> "NICMA MUSIC"
            else -> return
        }
        val targetDocumentId = artistPageDocumentId(ArtistPageBrand.Nicma, targetArtistName)
        collection.document(targetDocumentId).set(
            mapOf(
                "instagramURL" to page.instagramURL.trimmedOrNull(),
                "spotifyURL" to page.spotifyURL.trimmedOrNull(),
                "youtubeURL" to page.youtubeURL.trimmedOrNull(),
                "updatedAt" to Timestamp.now(),
            ),
            SetOptions.merge(),
        ).await()
    }

    private fun startObservation() {
        listener?.remove()
        listener = collection.addSnapshotListener { snapshot, error ->
            if (error != null) {
                _lastErrorMessage.value = error.message ?: "Artist-Seiten konnten nicht geladen werden."
                return@addSnapshotListener
            }

            val remotePages = snapshot?.documents.orEmpty()
                .mapNotNull { document ->
                    val brand = ArtistPageBrand.entries.firstOrNull {
                        it.rawValue == document.getString("brand")?.trim()?.lowercase()
                    } ?: return@mapNotNull null
                    val rawArtistName = document.getString("artistName")?.trim().orEmpty()
                    if (rawArtistName.isBlank()) return@mapNotNull null
                    val artistName = canonicalArtistNameForDocument(brand, document.id, rawArtistName)

                    ArtistPageUi(
                        slug = artistPageDocumentId(brand, artistName),
                        brand = brand,
                        artistName = artistName,
                        tagline = document.getString("tagline")?.trimmedOrNull(),
                        bio = document.getString("bio")?.trimmedOrNull(),
                        profileImageURL = document.getString("profileImageURL")?.trimmedOrNull(),
                        heroImageURL = document.getString("heroImageURL")?.trimmedOrNull(),
                        heroVideoURL = document.getString("heroVideoURL")?.trimmedOrNull(),
                        instagramURL = document.readArtistSocialUrl(
                            "instagram",
                            "instagramURL", "instagramUrl", "instagram", "instagramURLString",
                        ),
                        spotifyURL = document.readArtistSocialUrl(
                            "spotify",
                            "spotifyURL", "spotifyUrl", "spotify", "spotifyUrlString",
                        ),
                        youtubeURL = document.readArtistSocialUrl(
                            "youtube",
                            "youtubeURL", "youtubeUrl", "youtube", "youtubeURLString", "youTubeURL", "youTube",
                        ),
                        studioPriceList = ((document.get("studioPriceList") as? List<*>) ?: emptyList<Any>())
                            .mapNotNull { entry ->
                                val map = entry as? Map<*, *> ?: return@mapNotNull null
                                val title = (map["title"] as? String)?.trimmedOrNull() ?: return@mapNotNull null
                                val detail = (map["detail"] as? String)?.trimmedOrNull() ?: return@mapNotNull null
                                val price = (map["price"] as? String)?.trimmedOrNull() ?: return@mapNotNull null
                                StudioPriceItemUi(title = title, detail = detail, price = price)
                            },
                        editorUids = (document.get("editorUids") as? List<*>)?.mapNotNull { (it as? String)?.trimmedOrNull() }.orEmpty(),
                        createdAtEpochMillis = document.getTimestamp("createdAt")?.toDate()?.time ?: System.currentTimeMillis(),
                        updatedAtEpochMillis = document.getTimestamp("updatedAt")?.toDate()?.time ?: System.currentTimeMillis(),
                        isPlaceholder = false,
                    )
                }

            val merged = mergePages(remotePages)
            _pages.value = merged
            _lastErrorMessage.value = null
        }
    }

    private fun mergePages(remotePages: List<ArtistPageUi>): List<ArtistPageUi> {
        val pagesBySlug = linkedMapOf<String, ArtistPageUi>()
        remotePages.forEach { page ->
            val existing = pagesBySlug[page.slug]
            pagesBySlug[page.slug] = when {
                existing == null -> page
                existing.updatedAtEpochMillis != page.updatedAtEpochMillis ->
                    if (existing.updatedAtEpochMillis > page.updatedAtEpochMillis) existing else page
                existing.isPlaceholder != page.isPlaceholder ->
                    if (existing.isPlaceholder) page else existing
                else -> existing
            }
        }

        return pagesBySlug.values.sortedWith(
            compareBy<ArtistPageUi>({ it.brand.rawValue }, { it.artistName.lowercase() }),
        )
    }
}

/**
 * Manche aeltere oder manuell gepflegte Docs nutzen leicht abweichende Schluessel – gleich zu iOS-Video-Hub-Maps.
 */
private fun DocumentSnapshot.readFirstNonBlankString(vararg fieldNames: String): String? {
    for (name in fieldNames) {
        getString(name)?.trimmedOrNull()?.let { return it }
    }
    return null
}

/**
 * Wie [readFirstNonBlankString], zusaetzlich in `social` / `links` / `link` Maps
 * (typisch bei per Konsole gepflegten oder aelteren Shapes).
 */
private fun DocumentSnapshot.readArtistSocialUrl(
    groupKey: String,
    vararg topLevelNames: String,
): String? {
    readFirstNonBlankString(*topLevelNames)?.let { return it }
    when (val raw = get(groupKey)) {
        is String -> raw.trimmedOrNull()?.let { return it }
        is Map<*, *> -> {
            for (k in listOf("url", "link", "href", "u")) {
                (raw[k] as? String)?.trimmedOrNull()?.let { return it }
            }
        }
    }
    val nested = listOf("social", "links", "link", "${groupKey}URL")
    val nestedKeyCandidates = when (groupKey) {
        "instagram" -> listOf("instagram", "instagramURL", "instagramUrl", "ig")
        "spotify" -> listOf("spotify", "spotifyURL", "spotifyUrl", "s")
        "youtube" -> listOf("youtube", "youtubeURL", "youtubeUrl", "yt", "youTube", "youTubeURL")
        else -> emptyList()
    }
    for (mapName in nested) {
        val m = get(mapName) as? Map<*, *> ?: continue
        for (k in nestedKeyCandidates + "url" + "link" + "href") {
            (m[k] as? String)?.trimmedOrNull()?.let { return it }
        }
    }
    return null
}

private fun String?.trimmedOrNull(): String? {
    val trimmed = this?.trim().orEmpty()
    return trimmed.ifBlank { null }
}
